src/osm/GeometryHandler.cpp

```diff
@@ -61,137 +77,396 @@ osm2rdf::osm::GeometryHandler<W>::GeometryHandler(
+
   for (const auto& nodeRef : way.nodes()) {
     nodeIds.push_back(nodeRef.id());
   }
-  auto geom = way.geom();
-  if (_config.simplifyGeometries > 0) {
-    geom = simplifyGeometry(geom);
+  const auto& geom = simplifyGeometry(way.geom());
+
+  std::vector<osm2rdf::geometry::Box> boxes;
+
+  const size_t CHUNKSIZE = 25;
+
+  if (geom.size() >= 2 * CHUNKSIZE) {
+    size_t curSize = 0;
+    while (curSize < geom.size()) {
+      osm2rdf::geometry::Box box;
+
+      osm2rdf::geometry::Way tmp(
+          geom.begin() + curSize,
+          geom.begin() + std::min(curSize + CHUNKSIZE, geom.size()));
+
+      boost::geometry::envelope(tmp, box);
+      boxes.push_back(box);
+      curSize += CHUNKSIZE;
+    }
+  } else {
+    boxes.push_back(way.envelope());
   }
+
 #pragma omp critical(wayDataInsert)
   {
-    _oaWays << SpatialWayValue(way.envelope(), way.id(), geom, nodeIds);
+    _oaWays << SpatialWayValue(way.envelope(), way.id(), geom, nodeIds, boxes,
+                               pack(getBoxIds(geom, way.envelope())));
     _numWays++;
   }
 }
 
 // ____________________________________________________________________________
 template <typename W>
 template <typename G>
-G osm2rdf::osm::GeometryHandler<W>::simplifyGeometry(const G& g) {
-  G geom;
-  auto perimeter_or_length =
-      std::max(boost::geometry::perimeter(g), boost::geometry::length(g));
+G GeometryHandler<W>::simplifyGeometry(const G& geom) const {
+  G simplifiedGeom;
+
+  if (_config.simplifyGeometries == 0) {
+    // simple case, just remove colinear points
+    boost::geometry::simplify(geom, simplifiedGeom, 0);
+    return simplifiedGeom;
+  }
+
+  auto perimeterOrLength =
+      std::max(boost::geometry::perimeter(geom), boost::geometry::length(geom));
   do {
-    boost::geometry::simplify(
-        g, geom,
-        osm2rdf::osm::constants::BASE_SIMPLIFICATION_FACTOR *
-            perimeter_or_length * _config.simplifyGeometries);
-    perimeter_or_length /= 2;
-  } while (
-      (boost::geometry::is_empty(geom) || !boost::geometry::is_valid(geom)) &&
-      perimeter_or_length >=
-          osm2rdf::osm::constants::BASE_SIMPLIFICATION_FACTOR);
-  if (!boost::geometry::is_valid(geom)) {
-    return g;
+    boost::geometry::simplify(geom, simplifiedGeom,
+                              BASE_SIMPLIFICATION_FACTOR * perimeterOrLength *
+                                  _config.simplifyGeometries);
+    perimeterOrLength /= 2;
+  } while ((boost::geometry::is_empty(simplifiedGeom) ||
+            !boost::geometry::is_valid(simplifiedGeom)) &&
+           perimeterOrLength >= BASE_SIMPLIFICATION_FACTOR);
+  if (!boost::geometry::is_valid(simplifiedGeom)) {
+    return geom;
+  }
+  if (boost::geometry::is_empty(simplifiedGeom)) {
+    return geom;
+  }
+  return simplifiedGeom;
+}
+
+// ____________________________________________________________________________
+template <typename W>
+template <int MODE>
+bool GeometryHandler<W>::ioDouglasPeucker(
```

Comment:
`ioDouglasPeucker` -> `innerOuterDouglasPeucker`? If so it would be a few characters longer but more readable as `io` could be input/output as well.